### PR TITLE
[blog] Own Fixer post init

### PR DESCRIPTION
**Pull request by TomasVotruba**
Follow up post to: https://www.tomasvotruba.cz/blog/2017/07/17/how-to-write-custom-sniff-for-code-sniffer-3/

### CODE REVIEW

**Comment by TomasVotruba** [comment]
@SpacePossum Hey man, I wrote a post on how-to write fixer :). Could I ask for a short review?

**The post is focused on beginners**, so they can understand and focus on basics. I'll cover testing and more complex topics in future posts.

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+'''
+---
+
+When a coding standard tool finds over 1000 violations in our code is nice to know, but it doesn't save us any time and energy we need for <a href="@todo">a deep work</a>.
+
+### Find & Fix It
+
+That main difference of PHP-CS-Fixer to PHP_CodeSniffer is that **every Fixer has to fix issues it finds**. That's why there is no `LineLenghtFixer`, because fixing line length is difficult to automate.
+ 
+Personally I like PHP-CS-Fixer a bit more, because of more friendlier API, active community and openness to 3rd party packages. Working with tokens is still too painful to use practise, but I'll write about that later.
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: "too painful to use practise"?

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+{
+}'
+            ),
+        ]);
+    }        
+
+    // if you think this change code in any drastic  way, that changes behavior, return "true"
+    public function isRisky(): bool
+    {
+        return false;
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: This fixer is a good example of risky fixer. Change to true.

**Comment by SpacePossum** [comment]
I'm off for a few weeks so I can't do a review now, but nice work man :)

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md (hunk 1 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: Improved to be more specific

**Comment by TomasVotruba** [comment]
@enumag Thanks for feedback, feel free to comment anymore. I want to merge it now, because I have more posts to link one to another.

@SpacePossum Thanks for update!

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    public function getPriority(): int
+    {
+        return 0;
+    }        
+}
+```
+
+### 3. Subscribe the Fixer
+
+Now we get to more interesting parts. Method `isCandidate(Tokens $tokens): bool` is like a **subscriber**. It get all tokens of the file. We can check **more than one token and create more strict conditions** thanks to that:
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: It gets

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    // $tokens support array access - to get a token with some index, call $tokens[25]
+    $parentClassNameToken = $tokens[$parentClassNamePosition];
+    $parentClassName = $parentClassNameToken->getContent();
+
+    return $this->stringEndsWith($parentClassName, 'Exception');
+}
+
+private function stringEndsWith(string $name, string $needle): bool 
+{
+    return (substr($name, -strlen($needle)) === $needle);
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: useless brackets

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+}
+```
+
+Back to iteration! When this passes, we know **we have a class that extends an exception**.
+
+Do you know what we need to do now? You're right, **we have to check the name of it**. We can use another helper method: `getPrevMeaningfulToken()`.
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: we have to check it's name

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+    private function stringEndsWith(string $name, string $needle): bool 
+    {
+        return (substr($name, -strlen($needle)) === $needle);
+    }
+}
+```
+
+*How to run it?*
+
+With [EasyCodingStandard](https://github.com/Symplify/EasyCodingStandard) put the class to `easy-coding-standard.neon`:
```
Comment (enumag) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: 👎  In this article I'm far more interested in hot wo do it without ECS.

Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: Shall I link docs?

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    // this methods return the error message
+    // and it might include a sample code, that would fix it
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition('Exception classes should have suffix "Exception".', [
+            new CodeSample(
+                '<?php
+class SomeClass extends Exception
+{
+}'
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: Here I just used convention from PHP-CS-Fixer: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/Fixer/Basic/BracesFixer.php#L52

So it's not confusing while reading their code.

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md (hunk 6 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: Fixed'

File: source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    }        
+
+    // if you think this change code in any way, that changes behavior, return "true"
+    // changing a class name is such case
+    public function isRisky(): bool
+    {
+        return true;
+    }        
+
+    // in case you would like to support some specific files
```
Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: Only `*.php` are passed via internal iterator.
It is mostly `true`, only difference I found was this: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/dc04c34f7f2725a55b84d4b075a7d05586abb495/src/AbstractPsrAutoloadingFixer.php#L55

Comment (TomasVotruba) [inline source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md]: I've updated this section, to make it clear

**Comment by TomasVotruba** [comment]
Thanks for the feedback, I've integreated them to the post

**Pull request merged**